### ISSUE TITLE
Ideas needed: Workflows to keep edit files clear of foreign axioms

### ISSUE THREAD

**Issue by matentzn** [INCATools/ontology-development-kit#284]
Terminology: native axioms are axioms belonging to an ontology; foreign axioms belong to another ontology and are imported. Typically native axioms are:

- all axioms which mention at least one entity from the ontology namespace
- axiom annotations on other native axioms

Foreign axioms are everything else. Example: in MP
- MP1 sub BFO1 is native
- BFO1 sub BFO2 is foreign
- BFO1 rdfs:label "hello" is foreign

Exceptions could be xrefs:
- BFO1 xref: "MP:1" is foreign 
- BFO1 xref: MP1 should be foreign

No matter what I tried, foreign axioms keep creeping in all the edit files I am supporting, probably due to some weirdnesses in Protege, etc. These foreign axioms can confuse ROBOT validation as well; For example if you have a declaration of a GO class in your ontology, but no label; But more importantly, they may become outdated, refer to later obsoleted classes etc; so we really dont want them. But how do we manage their removal in a super simple way? 

Potential ideas:
1) I was thinking to add a command in ODK to simply print foreign axioms with SPARQL/ROBOT. That way people can simply choose to remove them manually. The downside is that people who only ever look at the ontology in Protege (there are many!) may have difficulties to really clear out axioms such as declarations or, god forbid, gcis.. 
2) Ther could be a ODK command that cleans the edit file for foreign axioms. The downside is that this would involve the ODK touching the edit file; I liked the fact that this never happened, because it protects the edit file from unintended changes due to some serialisation issue; but maybe that is over protective?

Any other idea? @cmungall @dosumis @balhoff

**Comment by jamesaoverton** [comment]
Yeah, @beckyjackson and I need something like this soonish. The immediate use case is to convert a released OWL file to something like a [base artifact](https://github.com/INCATools/ontology-development-kit/blob/master/docs/ReleaseArtefacts.md#release-artefact-1-base-required).

I could see this in ROBOT as `remove/filter --axioms foreign` or something. My first thought is to get the signature of each axiom as a set of IRI strings, and if none of those IRIs are in the current ontology's namespace (by comparing string prefixes) then this axiom is foreign. (Note that ROBOT isn't yet aware of "the current ontology's namespace".)

**Comment by matentzn** [comment]
So what we currently do is this:

1. We allow the users to specify "namespace of interest"; see [CL](https://github.com/INCATools/ontology-development-kit/blob/0b0e218da7c40a207d23e2c8a8df5bea3f7d3ffe/configs/cl-odk.yaml#L12) for an example.
2. Then we use this to create a [SPARQL query that selects all terms conforming to that spec](https://github.com/obophenotype/cell-ontology/blob/master/src/sparql/cl_terms.sparql)
3. We then use `filter --term-file $(SIMPLESEED) --select "annotations ontology anonymous self" --trim true --signature true` ([example](https://github.com/obophenotype/c-elegans-phenotype-ontology/blob/add8d79ac45c10c5cab4001d17d000556c84d16d/src/ontology/Makefile#L249))

To obtain only the "native" axioms. What do you think of that in general?

**Comment by balhoff** [comment]
@matentzn @beckyjackson what do you think about `external` rather than `foreign`?

**Comment by beckyjackson** [comment]
I just opened a WIP PR here: https://github.com/ontodev/robot/pull/570

@balhoff the current WIP uses `foreign` but I think `external` is a good alternative. What do others think?

**Comment by matentzn** [comment]
Happy with external/internal lingo!

**Comment by cmungall** [comment]
I prefer external/internal

I think we need to be clearer about the definitions. What about equivalence axioms between two NCs?

@matentzn can you talk more about the circumstances in which you see these creeping into the edit file? I have seen this with declarations, but not with non-declaration axioms

**Comment by matentzn** [comment]
The most important problem is that our current SOP requires the curator to create a new term as a subclass of thing; for example, let's say I want to add a logical definition involving GO:001, and GO:001 is not currently present in the GO module I am importing. The current SOP says: add GO:001 (the IRI) under Thing, and then use it in the logical definition. This creates two foreign axioms: The declaration, and the subclass of owl:Thing axiom, which need to be, periodically, cleaned up. I guess this problem will go away once everything is managed through DOSDP, but that is not going to be tomorrow, and this will only solve the problem for a handful of pattern-based phenotype ontologies..

I guess the problem could be solved merely through patience and training, and a failing pipeline if axioms are introduced accidentally (Protege bugs, or people accidentally drag-dropping a class in the external part of the hierarchy).. I dont mind. This is not a very high priority for me. And I am ok with being overruled on the ROBOT ticket as well; I just wanted to outline what my priorities are with regards to this functionality: I would like one quick method to divide the ontology into an external and internal part, given a set of "internal namespaces"; Around that, I can then easily design my QC framework.

**Comment by matentzn** [comment]
(Remember, many curators wont run the release pipeline to refresh imports, but they  often need a way to use external terms right away - hence the current way to do it with adding a declaration in Protege)

**Comment by gouttegd** [comment]
This does not seem relevant to me anymore? @matentzn ?

**Comment by matentzn** [comment]
The issue as significantly evolved but its still not sorted. Ideally you'd have something like:

```
clean-edit-file:
    robot remove --base-iri ... --axioms external -i $(SRC) -o $(SRC)
```

Or something bespoke. Aspecially cleaning up external declarations etc, but also in general.

I would leave this issue for now, but can put to backlog

**Comment by gouttegd** [comment]
Seems quite dangerous to me. I think it’s bold to assume that any “foreign axiom” (axioms pertaining solely to foreign entities) in the -edit file is necessarily the result of a mistake and can be safely removed.

For example, for quite a long time Uberon had been injecting axioms pertaining to RO relations. Yes, admittedly those axioms did not belong to Uberon, and in fact over time we upstreamed them to RO (or at least most of them? I don’t remember if we finished that job…), but until they could be moved to RO their presence in Uberon was _not_ a mistake.

**Comment by matentzn** [comment]
For sure. needs thinking, needs warnings, needs docs. But the reality is that "good" repos could benefit from these feature, especially for cleaning up the inevitably accumulating `Declarations()`.. Hence backlog.
